core/src/main/java/kafka/server/AssignmentsManager.java

```
@@ -0,0 +1,384 @@
+    private final int brokerId;
+    private final Supplier<Long> brokerEpochSupplier;
+    private final KafkaEventQueue eventQueue;
+
+    // These variables should only be mutated from the event loop thread
+    private Map<TopicIdPartition, AssignmentEvent> inflight = null;
+    private Map<TopicIdPartition, AssignmentEvent> pending = new HashMap<>();
+    private final ExponentialBackoff resendExponentialBackoff =
+            new ExponentialBackoff(100, 2, MAX_BACKOFF_INTERVAL_MS, 0.02);
+    private int failedAttempts = 0;
+
+    public AssignmentsManager(Time time,
+                              NodeToControllerChannelManager channelManager,
+                              int brokerId,
+                              Supplier<Long> brokerEpochSupplier) {
+        this.time = time;
+        this.channelManager = channelManager;
+        this.brokerId = brokerId;
+        this.brokerEpochSupplier = brokerEpochSupplier;
+        this.eventQueue = new KafkaEventQueue(time,
+                new LogContext("[AssignmentsManager id=" + brokerId + "]"),
+                "broker-" + brokerId + "-directory-assignments-manager-");
+    }
+
+    public void close() throws InterruptedException {
+        eventQueue.close();
+        channelManager.shutdown();
+    }
+
+    public void onAssignment(TopicIdPartition topicPartition, Uuid dirId) {
+        eventQueue.append(new AssignmentEvent(time.nanoseconds(), topicPartition, dirId));
+    }
+
+    // only for testing
+    void wakeup() {
+        eventQueue.wakeup();
+    }
+
+    /**
+     * Base class for all the events handled by {@link AssignmentsManager}.
+     */
+    private abstract static class Event implements EventQueue.Event {
+        /**
+         * Override the default behavior in
+         * {@link EventQueue.Event#handleException}
+         * which swallows the exception.
+         */
+        @Override
+        public void handleException(Throwable e) {
+            throw new RuntimeException(e);
```

Throwing an exception here isn't going to do anything useful. I mean, it will log an error message in `KafkaEventQueue.java`, but that's it.

If that's what you want to do to handle errors, I'd advise logging the error message here to make it clear that's what's going on.

---

This isn't great, but the default implementation is to ignore/swallow the exception. I guess it is better to have something in the logs rather than nothing at all. I guess it makes more sense for it to be logged here than in the underlying event queue, I'll make that change.